lib/OpenQA/Scheduler/Scheduler.pm

```diff
@@ -1208,23 +1208,36 @@ sub job_schedule_iso {
         asset_register(%$a);
     }
     my $noobsolete = delete $args{_NOOBSOLETEBUILD};
-    # ISOURL == ISO download. If the ISO already exists, skip the
-    # download step (below) entirely by leaving $isodlpath unset.
-    my $isodlpath;
-    if ($args{ISOURL}) {
+    # Any arg name ending in _URL is special: it tells us to download
+    # the file at that URL before running the job
+    my %downloads = ();
+    for my $arg (keys %args) {
+        next unless ($arg =~ /_URL$/);
         # As this comes in from an API call, URL will be URI-encoded
-        # This obviously creates a vuln if untrusted users can POST ISOs
-        $args{ISOURL} = url_unescape($args{ISOURL});
-        # set $args{ISO} to the URL filename if we only got ISOURL.
-        # This has to happen *before* _generate_jobs so the jobs have
-        # ISO set
-        if (!$args{ISO}) {
-            $args{ISO} = Mojo::URL->new($args{ISOURL})->path->parts->[-1];
+        # This obviously creates a vuln if untrusted users can POST
+        $args{$arg} = url_unescape($args{$arg});
+        my $url = $args{$arg};
+        # if $args{FOO_URL} is set but $args{FOO} is not, we will
+        # set $args{FOO} (the filename of the downloaded asset) to
+        # the URL filename. This has to happen *before*
+        # generate_jobs so the jobs have FOO set
+        my $short = substr($arg, 0, -4);
+        if (!$args{$short}) {
+            $args{$short} = Mojo::URL->new($url)->path->parts->[-1];
         }
-        # full path to download target location
-        my $fulliso = catfile($OpenQA::Utils::isodir, $args{ISO});
-        unless (-s $fulliso) {
-            $isodlpath = $fulliso;
+        # full path to download target location. We need to guess
+        # the asset type to know where to put it, using the same
+        # subroutine as parse_assets_from_settings
+        my $assettype = asset_type_from_setting($short);
+        # We're only going to allow downloading of asset types
+        die "_URL downloading only allowed for asset types! $short is not an asset type" unless ($assettype);
```

Comment:
I am not sure about dying here. I think this will kill the Scheduler DBUS handler and it will refuse to schedule new isos until rebooted. Sadly we don't have any text feedback from scheduler back to webapi, so I would just return empty array to signalize no job was created.